File: revenue/uiowa_rfq_18649_workbench/test_input_transport.py
"""App-source -> request bytes -> actual loopback HTTP parser regressions.

The recording adapter observes exactly what the unchanged server would supply to
its compiler. A separate test exercises the real compiler when the parent exists.
Node executes the actual app.js click handler with a controlled DOM, not a browser.
"""
from __future__ import annotations

import base64
import http.client
import importlib.util
import json
import shutil
import subprocess
import threading
import unittest
from pathlib import Path

HERE = Path(__file__).resolve().parent
SPEC = importlib.util.spec_from_file_location("workbench_transport_server", HERE / "server.py")
if SPEC is None or SPEC.loader is None:
    raise RuntimeError("workbench server loader unavailable")
server = importlib.util.module_from_spec(SPEC)
SPEC.loader.exec_module(server)


def browser_body(candidate: bytes, authority: bytes = b"{}") -> dict:
    node = shutil.which("node")
    if node is None:
        raise RuntimeError("Node.js is required to execute the actual workbench import path")
    payload = {"candidate_b64": base64.b64encode(candidate).decode("ascii"),
               "authority_b64": base64.b64encode(authority).decode("ascii")}
    result = subprocess.run(
        [node, str(HERE / "test_input_transport.js"), "--bridge"],
        input=json.dumps(payload), text=True, capture_output=True, timeout=20,
        check=True,
    )
    value = json.loads(result.stdout)
    value["body"] = (base64.b64decode(value["body_b64"], validate=True)
                     if value["body_b64"] is not None else None)
    return value


class RecordingAdapter:
    def __init__(self):
        self.calls = []

    def inspect(self, candidate, authority):
        self.calls.append((candidate, authority))
        return {"schema": "transport-test-recording-adapter", "mode": "UNTRUSTED_INSPECTION",
                "trust": {"current_evidence_review_authority": False},
                "candidate": candidate, "authority": authority}


class TransportHTTPTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.adapter = RecordingAdapter()
        cls.httpd = server.create_server(port=0, adapter=cls.adapter, static_root=HERE)
        cls.host, cls.port = cls.httpd.server_address[:2]
        cls.thread = threading.Thread(target=cls.httpd.serve_forever, daemon=True)
        cls.thread.start()

    @classmethod
    def tearDownClass(cls):
        cls.httpd.shutdown()
        cls.httpd.server_close()
        cls.thread.join(timeout=3)

    def setUp(self):
        self.adapter.calls.clear()

    def post(self, body: bytes, **extra_headers):
        conn = http.client.HTTPConnection(self.host, self.port, timeout=5)
        try:
            conn.request("POST", "/api/inspect", body=body, headers={
                "Content-Type": "application/json", "Host": f"127.0.0.1:{self.port}",
                "Origin": f"http://127.0.0.1:{self.port}", **extra_headers,
            })
            response = conn.getresponse()
            return response.status, dict(response.getheaders()), json.loads(response.read())
        finally:
            conn.close()

    def test_original_two_documents_reach_the_server(self):
        candidate = ' \r\n{"note":"caf\u00e9 \u6771\u4eac \U0001f642", "value":null}\t'.encode("utf-8")
        authority = b'\r\n{"sources": [], "label":"\\u0061"}\n'
        emitted = browser_body(candidate, authority)
        self.assertEqual(emitted["body"], b'{"candidate":' + candidate + b',"authority":' + authority + b'}')
        status, headers, _ = self.post(emitted["body"])
        self.assertEqual(status, 200)
        self.assertEqual(headers["Cache-Control"], "no-store")
        self.assertEqual(self.adapter.calls, [(json.loads(candidate), json.loads(authority))])

    def assert_duplicate_rejected(self, candidate, authority=b"{}"):
        emitted = browser_body(candidate, authority)
        self.assertEqual(emitted["requests"], 1)
        status, _, payload = self.post(emitted["body"])
        self.assertEqual(status, 400)
        self.assertIn("duplicate JSON key", payload["error"])
        self.assertEqual(self.adapter.calls, [])

    def test_candidate_duplicate_never_calls_the_compiler(self):
        self.assert_duplicate_rejected(b'{"count":1,"count":2}')

    def test_nested_duplicate_never_calls_the_compiler(self):
        self.assert_duplicate_rejected(b'{"rows":[{"count":1,"count":2}]}')

    def test_escaped_duplicate_never_calls_the_compiler(self):
        self.assert_duplicate_rejected(b'{"name":1,"n\\u0061me":2}')

    def test_authority_duplicate_never_calls_the_compiler(self):
        self.assert_duplicate_rejected(b'{}', b'{"sources":[],"sources":[1]}')

    def test_integer_outside_javascript_safe_range_arrives_exactly(self):
        candidate = b'{"count":9007199254740993,"negative":-9007199254740993}'
        emitted = browser_body(candidate)
        status, _, payload = self.post(emitted["body"])
        self.assertEqual(status, 200)
        self.assertEqual(self.adapter.calls[0][0]["count"], 9007199254740993)
        self.assertEqual(self.adapter.calls[0][0]["negative"], -9007199254740993)
        self.assertEqual(payload["report"]["candidate"]["count"], 9007199254740993)

    def test_lexical_number_forms_survive_to_parser_boundary(self):
        candidate = b'{"zero":-0,"decimal":1.2300,"exponent":1e+30}'
        emitted = browser_body(candidate)
        self.assertIn(candidate, emitted["body"])
        self.assertEqual(self.post(emitted["body"])[0], 200)
        self.assertEqual(self.adapter.calls[0][0], json.loads(candidate))

    def test_invalid_utf8_stops_before_http(self):
        emitted = browser_body(b'{"note":"\xc3("}')
        self.assertIsNone(emitted["body"])
        self.assertEqual(emitted["requests"], 0)
        self.assertIn("valid UTF-8", emitted["error"])
        self.assertEqual(self.adapter.calls, [])

    def test_bom_is_not_silently_removed(self):
        emitted = browser_body(b'\xef\xbb\xbf{}')
        self.assertIsNone(emitted["body"])
        self.assertIn("valid JSON", emitted["error"])

    def test_literal_replacement_character_remains_valid_data(self):
        candidate = '{"note":"\ufffd"}'.encode("utf-8")
        emitted = browser_body(candidate)
        self.assertEqual(self.post(emitted["body"])[0], 200)
        self.assertEqual(self.adapter.calls[0][0]["note"], "\ufffd")

    def test_combined_byte_limit_stops_before_http(self):
        candidate = b'{"x":"' + b'x' * (1024*1024-8) + b'"}'
        emitted = browser_body(candidate, candidate)
        self.assertIsNone(emitted["body"])
        self.assertIn("2 MiB", emitted["error"])
        self.assertEqual(self.adapter.calls, [])

    def test_exactly_full_body_is_accepted_by_unchanged_server(self):
        overhead = len(b'{"candidate":,"authority":}')
        candidate = b'{"x":"' + b'x' * (1024*1024-8) + b'"}'
        authority = b'{"x":"' + b'x' * (1024*1024-overhead-8) + b'"}'
        emitted = browser_body(candidate, authority)
        self.assertEqual(len(emitted["body"]), server.MAX_BODY_BYTES)
        self.assertEqual(self.post(emitted["body"])[0], 200)
        self.assertEqual(len(self.adapter.calls), 1)

    def test_same_origin_boundary_is_unchanged(self):
        emitted = browser_body(b'{}')
        self.assertEqual(self.post(emitted["body"], Origin="http://example.invalid")[0], 403)
        self.assertEqual(self.adapter.calls, [])

    def test_no_new_trusted_root_input_is_permitted(self):
        emitted = browser_body(b'{}')
        modified = emitted["body"][:-1] + b',"trusted_authority_sha256":"fake"}'
        self.assertEqual(self.post(modified)[0], 400)
        self.assertEqual(self.adapter.calls, [])


class RealCompilerTransportTests(unittest.TestCase):
    def test_existing_parent_receipt_matches_direct_inspection(self):
        parent = HERE.parent / "uiowa_rfq_18649_workshare"
        if not parent.is_dir():
            self.skipTest("parent compiler absent in isolated transport checkout; no compiler-pass claim")
        candidate = (parent / "fixtures" / "synthetic_packet.json").read_bytes()
        authority = (parent / "fixtures" / "synthetic_authority.json").read_bytes()
        adapter = server.CompilerAdapter()
        expected = adapter.inspect(server.loads_strict_json(candidate), server.loads_strict_json(authority))
        emitted = browser_body(candidate, authority)
        self.assertIsNotNone(emitted["body"], emitted["error"])
        httpd = server.create_server(port=0, adapter=adapter, static_root=HERE)
        thread = threading.Thread(target=httpd.serve_forever, daemon=True)
        thread.start()
        conn = http.client.HTTPConnection(*httpd.server_address, timeout=10)
        try:
            host = f"127.0.0.1:{httpd.server_address[1]}"
            conn.request("POST", "/api/inspect", body=emitted["body"], headers={
                "Host": host, "Origin": f"http://{host}", "Content-Type": "application/json"})
            response = conn.getresponse()
            payload = json.loads(response.read())
            self.assertEqual(response.status, 200, payload)
            self.assertEqual(payload["report"], expected)
            self.assertEqual(payload["report"]["mode"], "UNTRUSTED_INSPECTION")
            self.assertFalse(payload["report"]["trust"]["current_evidence_review_authority"])
        finally:
            conn.close()
            httpd.shutdown()
            httpd.server_close()
            thread.join(timeout=3)


if __name__ == "__main__":
    unittest.main()

File: revenue/uiowa_rfq_18649_workbench/test_input_transport.js
"use strict";
// Executes the actual app.js file and its registered import handler. The DOM and
// response are controlled test doubles; this is not rendered-browser acceptance.
const assert = require("node:assert/strict");
const fs = require("node:fs");
const path = require("node:path");
const vm = require("node:vm");

function element() {
  return { value: "", disabled: false, textContent: "", files: [], dataset: {},
    children: [], listeners: {},
    addEventListener(name, callback) { this.listeners[name] = callback; },
    replaceChildren(...children) { this.children = children; },
    append(...children) { this.children.push(...children); },
    setAttribute() {}, focus() {}, click() {}, remove() {},
    get childElementCount() { return this.children.length; } };
}
function file(bytes, size) {
  const blob = new Blob([bytes]);
  return { size: size ?? blob.size, arrayBuffer: () => blob.arrayBuffer(), text: () => blob.text() };
}
function makeApp(candidate, authority, options = {}) {
  const elements = new Map();
  const get = id => {
    if (!elements.has(id)) elements.set(id, element());
    return elements.get(id);
  };
  const calls = [];
  const context = vm.createContext({
    document: { getElementById: get, createElement: element, body: element() },
    Option: function(label, value) { return { textContent: label, value }; },
    location: { search: "" }, URLSearchParams, TextDecoder, TextEncoder, Blob, URL,
    console,
    fetch: async (url, request) => {
      calls.push({ url, ...request });
      return { ok: false, status: 400, json: async () => ({ error: "test endpoint rejected input" }) };
    }
  });
  const appPath = options.appPath || process.env.WORKBENCH_APP || path.join(__dirname, "app.js");
  // Pending shared handoff helpers are loaded when present, preserving a single
  // test entry point when the adjacent restore/Markdown work is composed.
  for (const helper of ["handoff.js", "draft_import.js"]) {
    const helperPath = path.join(path.dirname(appPath), helper);
    if (fs.existsSync(helperPath)) vm.runInContext(fs.readFileSync(helperPath, "utf8"), context, { filename: helper });
  }
  vm.runInContext(fs.readFileSync(appPath, "utf8"), context, { filename: appPath });
  if (candidate !== null) get("candidateFile").files = [file(candidate, options.candidateSize)];
  if (authority !== null) get("authorityFile").files = [file(authority)];
  return { context, elements, get, calls, async inspect() { await get("inspectBtn").listeners.click(); } };
}
async function transmit(candidate, authority = "{}", options = {}) {
  const app = makeApp(candidate, authority, options);
  await app.inspect();
  return app;
}
function emitted(app) {
  assert.equal(app.calls.length, 1, app.get("error").textContent);
  return app.calls[0].body;
}
const tests = [];
function test(name, fn) { tests.push([name, fn]); }

test("actual handler sends the original whitespace, CRLF and Unicode", async () => {
  const candidate = ' \r\n{ "note" : "caf\u00e9 \u6771\u4eac \ud83d\ude42", "empty":"", "missing": null }\t\n';
  const authority = '\n{ "sources" : [] }\r\n';
  const app = await transmit(candidate, authority);
  assert.equal(emitted(app), `{"candidate":${candidate},"authority":${authority}}`);
});
test("unsafe integers are not rounded by browser transport", async () => {
  const candidate = '{"count":9007199254740993,"negative":-9007199254740993}';
  assert.equal(emitted(await transmit(candidate)), `{"candidate":${candidate},"authority":{}}`);
});
test("number spelling survives without a browser numeric conversion", async () => {
  const candidate = '{"a":-0,"b":1.2300,"c":1e+30,"d":1e400,"e":1e-400}';
  assert.equal(emitted(await transmit(candidate)), `{"candidate":${candidate},"authority":{}}`);
});
for (const [label, candidate, authority] of [
  ["candidate top-level", '{"same":1,"same":2}', '{}'],
  ["nested", '{"a":[{"same":1,"same":2}]}', '{}'],
  ["escaped key", '{"same":1,"s\\u0061me":2}', '{}'],
  ["authority", '{}', '{"sources":[],"sources":[1]}']
]) test(`duplicate ${label} members reach the strict server unchanged`, async () => {
  assert.equal(emitted(await transmit(candidate, authority)), `{"candidate":${candidate},"authority":${authority}}`);
});
test("string escapes, literal replacement characters and key order survive", async () => {
  const candidate = '{"z":"\\u0061","a":"\ufffd","v":"line\\nnext","q":"\\\""}';
  assert.equal(emitted(await transmit(candidate)), `{"candidate":${candidate},"authority":{}}`);
});
for (const [label, bytes] of [
  ["invalid continuation", Buffer.from([123,34,120,34,58,34,0xc3,0x28,34,125])],
  ["truncated sequence", Buffer.from([123,34,120,34,58,34,0xe2,0x82,34,125])],
  ["encoded surrogate", Buffer.from([123,34,120,34,58,34,0xed,0xa0,0x80,34,125])]
]) test(`${label} UTF-8 fails before fetch instead of replacing evidence`, async () => {
  const app = await transmit(bytes);
  assert.equal(app.calls.length, 0);
  assert.match(app.get("error").textContent, /valid UTF-8/);
});
test("UTF-8 BOM is rejected rather than silently stripped", async () => {
  const app = await transmit(Buffer.concat([Buffer.from([0xef,0xbb,0xbf]), Buffer.from('{}')]));
  assert.equal(app.calls.length, 0);
  assert.match(app.get("error").textContent, /valid JSON/);
});
for (const value of ["", "null", "[]", "true", '"object"', "42", '{} ,"authority":{}', '{}{}', '{"x":NaN}']) {
  test(`invalid shape or syntax ${JSON.stringify(value)} cannot alter the envelope`, async () => {
    const app = await transmit(value);
    assert.equal(app.calls.length, 0);
    assert.match(app.get("error").textContent, /JSON/);
    assert.equal(app.get("exportBtn").disabled, true);
    assert.equal(app.get("inspectBtn").disabled, false);
  });
}
test("missing candidate is a visible error without a request", async () => {
  const app = await transmit(null);
  assert.equal(app.calls.length, 0);
  assert.match(app.get("error").textContent, /Candidate file is required/);
});
test("invalid authority is also rejected before fetch", async () => {
  const app = await transmit('{}', 'null');
  assert.equal(app.calls.length, 0);
  assert.match(app.get("error").textContent, /Authority must be a JSON object/);
});
test("one file over the byte cap is rejected", async () => {
  const app = await transmit('{"a":"' + 'x'.repeat(1024*1024) + '"}');
  assert.equal(app.calls.length, 0);
  assert.match(app.get("error").textContent, /1 MiB/);
});
test("decoded bytes are independently bounded even with a wrong size declaration", async () => {
  const app = await transmit('{"a":"' + 'x'.repeat(1024*1024) + '"}', '{}', { candidateSize: 1 });
  assert.equal(app.calls.length, 0);
  assert.match(app.get("error").textContent, /1 MiB/);
});
function paddedObject(bytes) { return '{"x":"' + 'x'.repeat(bytes-8) + '"}'; }
test("two full-size files cannot overflow the server envelope limit", async () => {
  const one = paddedObject(1024*1024);
  const app = await transmit(one, one);
  assert.equal(app.calls.length, 0);
  assert.match(app.get("error").textContent, /Combined.*2 MiB/);
});
test("an exactly full server envelope is accepted by the byte check", async () => {
  const overhead = Buffer.byteLength('{"candidate":,"authority":}');
  const app = await transmit(paddedObject(1024*1024), paddedObject(1024*1024-overhead));
  assert.equal(Buffer.byteLength(emitted(app)), 2*1024*1024);
});
test("multibyte data is bounded in bytes rather than character count", async () => {
  const app = await transmit('{"x":"' + '\u6771'.repeat(350000) + '"}');
  assert.equal(app.calls.length, 0);
  assert.match(app.get("error").textContent, /1 MiB/);
});
test("request contract and cache/origin settings remain unchanged", async () => {
  const app = await transmit('{}');
  emitted(app);
  const request = app.calls[0];
  assert.equal(request.url, '/api/inspect');
  assert.equal(request.method, 'POST');
  assert.equal(request.headers['Content-Type'], 'application/json');
  assert.equal(request.credentials, 'same-origin');
  assert.equal(request.cache, 'no-store');
});
test("a failed replacement clears prior report, notes and export state", async () => {
  const app = makeApp(Buffer.from([0xff]), '{}');
  vm.runInContext('installReport(syntheticReport()); selectCell(keyFor(state.cells[0])); state.notes.set(state.selectedKey, "old note");', app.context);
  assert.equal(app.get('exportBtn').disabled, false);
  await app.inspect();
  assert.equal(vm.runInContext('state.report', app.context), null);
  assert.equal(vm.runInContext('state.notes.size', app.context), 0);
  assert.equal(app.get('exportBtn').disabled, true);
  assert.equal(app.get('inspectBtn').disabled, false);
  assert.equal(app.calls.length, 0);
});

async function main() {
  if (process.argv.includes('--bridge')) {
    const input = JSON.parse(fs.readFileSync(0, 'utf8'));
    const app = await transmit(Buffer.from(input.candidate_b64,'base64'), Buffer.from(input.authority_b64,'base64'));
    process.stdout.write(JSON.stringify({
      body_b64: app.calls.length ? Buffer.from(app.calls[0].body, 'utf8').toString('base64') : null,
      requests: app.calls.length, error: app.get('error').textContent
    }));
    return;
  }
  let failures = 0;
  for (const [name, fn] of tests) {
    try { await fn(); console.log(`PASS ${name}`); }
    catch (err) { failures++; console.error(`FAIL ${name}\n${err.stack}`); }
  }
  console.log(`INPUT_TRANSPORT_NODE tests=${tests.length} passed=${tests.length-failures} failed=${failures}`);
  process.exitCode = failures ? 1 : 0;
}
main().catch(err => { console.error(err); process.exitCode = 1; });
